dali/operators/generic/roi_random_crop.cc

```diff
@@ -0,0 +1,207 @@
+If the ROI is smaller than the cropping window, the whole ROI should be contained in the cropping window.
+
+If an input shape (``in_shape``) is given, the resulting cropping window is selected to be within the
+bounds of that input shape. Alternatively, the input data subject to cropping can be fed directly to
+the operator.
+
+If no input shape is provided, the resulting cropping window is unbounded, potentially resulting in out 
+of bounds cropping.
+
+The cropping window dimensions should be explicitly provided (``crop_shape``), and the ROI should be
+either specified with ``roi_start``/``roi_end`` or ``roi_start``/``roi_shape``.
+
+The operator produces an output representing the cropping window start coordinates.
+)code")
+    .AddArg("crop_shape",
+      R"code(Cropping window dimensions.)code", DALI_INT_VEC, true)
+    .AddArg("roi_start",
+      R"code(ROI start coordinates.)code", DALI_INT_VEC, true)
+    .AddOptionalArg<std::vector<int>>("roi_end",
+      R"code(ROI end coordinates.
+
+Note: Using ``roi_end`` is mutually exclusive with ``roi_shape``.)code", nullptr, true)
+    .AddOptionalArg<std::vector<int>>("roi_shape",
+      R"code(ROI shape.
+
+Note: Using ``roi_shape`` is mutually exclusive with ``roi_end``.)code", nullptr, true)
+    .AddOptionalArg<std::vector<int>>("in_shape",
+      R"code(Shape of the input data.
+
+If provided, the cropping window start will be selected so that the cropping window is within the
+bounds of the input.
+
+Note: Providing ``in_shape`` is incompatible with feeding the input data directly as a positional input.
+)code", nullptr, true)
+    .NumInput(0, 1)
+    .NumOutput(1);
+
+class ROIRandomCropCPU : public Operator<CPUBackend> {
+ public:
+  explicit ROIRandomCropCPU(const OpSpec &spec);
+  bool CanInferOutputs() const override { return true; }
+  bool SetupImpl(std::vector<OutputDesc> &output_desc, const workspace_t<CPUBackend> &ws) override;
+  void RunImpl(workspace_t<CPUBackend> &ws) override;
+
+ private:
+  BatchRNG<std::mt19937> rngs_;
+
+  ArgValue<int, 1> roi_start_;
+  ArgValue<int, 1> roi_end_;
+  ArgValue<int, 1> roi_shape_;
+  ArgValue<int, 1> crop_shape_;
+  ArgValue<int, 1> in_shape_arg_;
+
+  TensorListShape<> in_shape_;
+
+  USE_OPERATOR_MEMBERS();
+};
+
+ROIRandomCropCPU::ROIRandomCropCPU(const OpSpec &spec)
+    : Operator<CPUBackend>(spec),
+      rngs_(spec.GetArgument<int64_t>("seed"), spec.GetArgument<int64_t>("max_batch_size")),
+      roi_start_("roi_start", spec),
+      roi_end_("roi_end", spec),
+      roi_shape_("roi_shape", spec),
+      crop_shape_("crop_shape", spec),
+      in_shape_arg_("in_shape", spec) {
+  if ((roi_end_.IsDefined() + roi_shape_.IsDefined()) != 1)
+    DALI_FAIL("Either ROI end or ROI shape should be defined, but not both");
+}
+
+bool ROIRandomCropCPU::SetupImpl(std::vector<OutputDesc> &output_desc,
+                                 const workspace_t<CPUBackend> &ws) {
+  int nsamples = spec_.HasTensorArgument("crop_shape") ?
```

Comment:
Maybe we should enforce that either `crop_shape` or input provided. Now if we won't provide anything it will crash (I guess).